### ISSUE TITLE
Crash when setting NFY:ledmode

### ISSUE THREAD

**Issue by nanofury** [luke-jr/bfgminer#426]
Specifying --set-device NFY:ledmode=alternating at the command line crashes bfgminer

From the log file:

 [2014-04-07 22:46:45] Pushing work 16 from pool 0 to hash queue
 [2014-04-07 22:46:45] Selecting pool 0 for work
 [2014-04-07 22:46:45] Generated stratum header 000000026d78db690f874caeb17bd3c66b8b31842aee5ed8282be3d30000000000000000386abb19eff84a85c5c15092558a9234f85145c41274a9f4cd0521d1dba1be8e53438cc41900b3aa00000000
 [2014-04-07 22:46:45] Work job_id 1396935876 243657 nonce2 08000000
 [2014-04-07 22:46:45] Generated target 0000000010000000000000000000000000000000000000000000000000000000
 [2014-04-07 22:46:45] Generated stratum work
 [2014-04-07 22:46:45] Pushing work 18 from pool 0 to hash queue
 [2014-04-07 22:46:45] mcp2210_spi_transfer: SPI transfer rejected temporarily (60 bytes remaining)
 [2014-04-07 22:46:45] mcp2210_spi_transfer: SPI transfer rejected temporarily (60 bytes remaining)
 [2014-04-07 22:46:45] mcp2210_spi_transfer: SPI transfer rejected temporarily (60 bytes remaining)
 [2014-04-07 22:46:45] mcp2210_spi_transfer: SPI transfer rejected temporarily (60 bytes remaining)
 [2014-04-07 22:46:45] mcp2210_spi_transfer: SPI transfer rejected temporarily (60 bytes remaining)
 [2014-04-07 22:46:46] mcp2210_spi_transfer: SPI transfer rejected temporarily (60 bytes remaining)
 [2014-04-07 22:46:46] mcp2210_spi_transfer: SPI transfer rejected temporarily (60 bytes remaining)
 [2014-04-07 22:46:46] NFY 0a: bitfury_init_oldbuf: Differ at 1
 [2014-04-07 22:46:46] NFY 0a: bitfury_init_oldbuf: Second differ at 2; trying again
 [2014-04-07 22:46:46] mcp2210_spi_transfer: SPI transfer rejected temporarily (60 bytes remaining)
 [2014-04-07 22:46:46] NFY 0a: bitfury_init_oldbuf: Differ at 1
 [2014-04-07 22:46:46] NFY 0a: bitfury_init_oldbuf: Second differ at 2; trying again
 [2014-04-07 22:46:46] mcp2210_spi_transfer: SPI transfer rejected temporarily (60 bytes remaining)
 [2014-04-07 22:46:46] NFY 0a: bitfury_init_oldbuf: Differ at 1
 [2014-04-07 22:46:46] NFY 0a: bitfury_init_oldbuf: Second differ at 2; trying again
 [2014-04-07 22:46:46] mcp2210_spi_transfer: SPI transfer rejected temporarily (60 bytes remaining)
 [2014-04-07 22:46:46] NFY 0a: bitfury_init_oldbuf: Differ at 1
 [2014-04-07 22:46:46] NFY 0a: Init a0803fd3>df899c19 df899d6a 5642780c e9444010 a0803fd3 df89a20d ff70600c a0803fd3 df89a5e0 df89a699 a0803fd3 df89a907 df89aa44 df8a38b0 ee642010 (job=00000000)
 [2014-04-07 22:46:46] mcp2210_spi_transfer: SPI transfer rejected temporarily (60 bytes remaining)
 [2014-04-07 22:46:46] mcp2210_spi_transfer: SPI transfer rejected temporarily (60 bytes remaining)
 [2014-04-07 22:46:46] mcp2210_spi_transfer: SPI transfer rejected temporarily (60 bytes remaining)
 [2014-04-07 22:46:46] mcp2210_spi_transfer: SPI transfer rejected temporarily (60 bytes remaining)
 [2014-04-07 22:46:46] mcp2210_spi_transfer: SPI transfer rejected temporarily (60 bytes remaining)
 [2014-04-07 22:46:47] mcp2210_spi_transfer: SPI transfer rejected temporarily (60 bytes remaining)
 [2014-04-07 22:46:47] mcp2210_spi_transfer: SPI transfer rejected temporarily (60 bytes remaining)
 [2014-04-07 22:46:47] NFY 0b: bitfury_init_oldbuf: Differ at 1
 [2014-04-07 22:46:47] NFY 0b: bitfury_init_oldbuf: Second differ at 2; trying again
 [2014-04-07 22:46:47] mcp2210_spi_transfer: SPI transfer rejected temporarily (60 bytes remaining)
 [2014-04-07 22:46:47] NFY 0b: bitfury_init_oldbuf: Differ at 1
 [2014-04-07 22:46:47] NFY 0b: bitfury_init_oldbuf: Second differ at 2; trying again
 [2014-04-07 22:46:47] mcp2210_spi_transfer: SPI transfer rejected temporarily (60 bytes remaining)
 [2014-04-07 22:46:47] NFY 0b: bitfury_init_oldbuf: Differ at 1
 [2014-04-07 22:46:47] NFY 0b: bitfury_init_oldbuf: Second differ at 2; trying again
 [2014-04-07 22:46:47] mcp2210_spi_transfer: SPI transfer rejected temporarily (60 bytes remaining)
 [2014-04-07 22:46:47] NFY 0b: bitfury_init_oldbuf: Differ at 1
 [2014-04-07 22:46:47] NFY 0b: Init f144e61f>df80efd8 df80f115 df80f265 df80f3b7 f04141e0 f144e61f df80f72d df80f8f8 f144e61f df80fb30 df80fc32 f144e61f df81be88 df819471 784161e0 (job=00000000)
 [2014-04-07 22:46:47] NFY 0a: cgpu_set_defaults: Matched with set default: NFY:osc6_bits=56
 [2014-04-07 22:46:47] NFY 0a: Applied rule NFY:osc6_bits=56
 [2014-04-07 22:46:47] NFY 0a: cgpu_set_defaults: Matched with set default: NFY:baud=500000
 [2014-04-07 22:46:47] NFY 0a: Applied rule NFY:baud=500000
 [2014-04-07 22:46:47] NFY 0a: cgpu_set_defaults: Matched with set default: NFY:ledmode=alternating
 [2014-04-07 22:46:47] NFY 0a: Applied rule NFY:ledmode=alternating
 [2014-04-07 22:46:47] NFY 0b: cgpu_set_defaults: Matched with set default: NFY:osc6_bits=56
 [2014-04-07 22:46:47] NFY 0b: Applied rule NFY:osc6_bits=56
 [2014-04-07 22:46:47] NFY 0b: cgpu_set_defaults: Matched with set default: NFY:baud=500000
 [2014-04-07 22:46:47] NFY 0b: Applied rule NFY:baud=500000
 [2014-04-07 22:46:47] NFY 0b: cgpu_set_defaults: Matched with set default: NFY:ledmode=alternating
 [2014-04-07 22:46:47] mcp2210_spi_transfer: SPI transfer rejected temporarily (60 bytes remaining)
0x4992a7 : C:\BTC\bfgminer-nanofury-win32\bfgminer-3-10-NF2\8fdaaffcdd662acf927f7e37515133a01e475579-win32\bfgminer.exe : /home/luke-jr/public_html/tmp/code/webisect/wr/checkout-U0C4-X8AAAEAAH@AlmQAAAAV/make-release-tmp/8fdaaffcdd662acf927f7e37515133a01e475579-tmp/driver-nanofury.c (480) : in function (nanofury_set_ledmode) 
0x42ef91 : C:\BTC\bfgminer-nanofury-win32\bfgminer-3-10-NF2\8fdaaffcdd662acf927f7e37515133a01e475579-win32\bfgminer.exe : /home/luke-jr/public_html/tmp/code/webisect/wr/checkout-U0C4-X8AAAEAAH@AlmQAAAAV/make-release-tmp/8fdaaffcdd662acf927f7e37515133a01e475579-tmp/deviceapi.c (931) : in function (_proc_set_device) 
0x42f0b8 : C:\BTC\bfgminer-nanofury-win32\bfgminer-3-10-NF2\8fdaaffcdd662acf927f7e37515133a01e475579-win32\bfgminer.exe : /home/luke-jr/public_html/tmp/code/webisect/wr/checkout-U0C4-X8AAAEAAH@AlmQAAAAV/make-release-tmp/8fdaaffcdd662acf927f7e37515133a01e475579-tmp/deviceapi.c (955) : in function (__proc_set_device) 
0x42f0e7 : C:\BTC\bfgminer-nanofury-win32\bfgminer-3-10-NF2\8fdaaffcdd662acf927f7e37515133a01e475579-win32\bfgminer.exe : /home/luke-jr/public_html/tmp/code/webisect/wr/checkout-U0C4-X8AAAEAAH@AlmQAAAAV/make-release-tmp/8fdaaffcdd662acf927f7e37515133a01e475579-tmp/deviceapi.c (959) : in function (proc_set_device) 
0x421c4b : C:\BTC\bfgminer-nanofury-win32\bfgminer-3-10-NF2\8fdaaffcdd662acf927f7e37515133a01e475579-win32\bfgminer.exe : /home/luke-jr/public_html/tmp/code/webisect/wr/checkout-U0C4-X8AAAEAAH@AlmQAAAAV/make-release-tmp/8fdaaffcdd662acf927f7e37515133a01e475579-tmp/miner.c (9773) : in function (cgpu_set_defaults) 
0x4222a2 : C:\BTC\bfgminer-nanofury-win32\bfgminer-3-10-NF2\8fdaaffcdd662acf927f7e37515133a01e475579-win32\bfgminer.exe : /home/luke-jr/public_html/tmp/code/webisect/wr/checkout-U0C4-X8AAAEAAH@AlmQAAAAV/make-release-tmp/8fdaaffcdd662acf927f7e37515133a01e475579-tmp/miner.c (9930) : in function (bfg_watchdog) 
0x42223b : C:\BTC\bfgminer-nanofury-win32\bfgminer-3-10-NF2\8fdaaffcdd662acf927f7e37515133a01e475579-win32\bfgminer.exe : /home/luke-jr/public_html/tmp/code/webisect/wr/checkout-U0C4-X8AAAEAAH@AlmQAAAAV/make-release-tmp/8fdaaffcdd662acf927f7e37515133a01e475579-tmp/miner.c (9910) : in function (watchdog_thread) 
Failed to init bfd from (C:\BTC\bfgminer-nanofury-win32\bfgminer-3-10-NF2\8fdaaffcdd662acf927f7e37515133a01e475579-win32\pthreadGC2.dll)
0x6248bbcb : C:\BTC\bfgminer-nanofury-win32\bfgminer-3-10-NF2\8fdaaffcdd662acf927f7e37515133a01e475579-win32\pthreadGC2.dll : pthread_create 
Failed to init bfd from (C:\Windows\system32\msvcrt.dll)
0x771b2599 : C:\Windows\system32\msvcrt.dll : wcstombs 
Failed to init bfd from (C:\Windows\system32\msvcrt.dll)
0x771b26b3 : C:\Windows\system32\msvcrt.dll : beginthreadex 
Failed to init bfd from (C:\Windows\system32\kernel32.dll)
0x7600d2e9 : C:\Windows\system32\kernel32.dll : BaseThreadInitThunk 
Failed to init bfd from (C:\Windows\system32\ntdll.dll)
0x77471603 : C:\Windows\system32\ntdll.dll : RtlInitializeExceptionChain 
Failed to init bfd from (C:\Windows\system32\ntdll.dll)
0x774715d6 : C:\Windows\system32\ntdll.dll : RtlInitializeExceptionChain 

If I'm reading the log file correctly - it fails at the line after the IF statement:

```
if (!strcasecmp(setting, "alternating"))
    state->ledalternating = true;
```

and it seems to be trying to set the state on a per-chip basis (as opposed to a per-device one).

I've noticed a similar issue (not yet reported as a separate ticket) where if I try to do IDENTIFY on any other chip than the first one it also crashes.

That kind of crash produces this in the log file:

 [2014-04-07 22:59:59] Generated stratum work
 [2014-04-07 22:59:59] Pushing work 7022 from pool 0 to hash queue
0x4991e6 : C:\BTC\bfgminer-nanofury-win32\bfgminer-3-10-NF2\8fdaaffcdd662acf927f7e37515133a01e475579-win32\bfgminer.exe : /home/luke-jr/public_html/tmp/code/webisect/wr/checkout-U0C4-X8AAAEAAH@AlmQAAAAV/make-release-tmp/8fdaaffcdd662acf927f7e37515133a01e475579-tmp/driver-nanofury.c (451) : in function (nanofury_identify) 
0x4196be : C:\BTC\bfgminer-nanofury-win32\bfgminer-3-10-NF2\8fdaaffcdd662acf927f7e37515133a01e475579-win32\bfgminer.exe : /home/luke-jr/public_html/tmp/code/webisect/wr/checkout-U0C4-X8AAAEAAH@AlmQAAAAV/make-release-tmp/8fdaaffcdd662acf927f7e37515133a01e475579-tmp/miner.c (7310) : in function (manage_device) 
0x419b27 : C:\BTC\bfgminer-nanofury-win32\bfgminer-3-10-NF2\8fdaaffcdd662acf927f7e37515133a01e475579-win32\bfgminer.exe : /home/luke-jr/public_html/tmp/code/webisect/wr/checkout-U0C4-X8AAAEAAH@AlmQAAAAV/make-release-tmp/8fdaaffcdd662acf927f7e37515133a01e475579-tmp/miner.c (7483) : in function (input_thread) 
Failed to init bfd from (C:\BTC\bfgminer-nanofury-win32\bfgminer-3-10-NF2\8fdaaffcdd662acf927f7e37515133a01e475579-win32\pthreadGC2.dll)
0x6248bbcb : C:\BTC\bfgminer-nanofury-win32\bfgminer-3-10-NF2\8fdaaffcdd662acf927f7e37515133a01e475579-win32\pthreadGC2.dll : pthread_create 
Failed to init bfd from (C:\Windows\system32\msvcrt.dll)
0x771b2599 : C:\Windows\system32\msvcrt.dll : wcstombs 
Failed to init bfd from (C:\Windows\system32\msvcrt.dll)
0x771b26b3 : C:\Windows\system32\msvcrt.dll : beginthreadex 
Failed to init bfd from (C:\Windows\system32\kernel32.dll)
0x7600d2e9 : C:\Windows\system32\kernel32.dll : BaseThreadInitThunk 
Failed to init bfd from (C:\Windows\system32\ntdll.dll)
0x77471603 : C:\Windows\system32\ntdll.dll : RtlInitializeExceptionChain 
Failed to init bfd from (C:\Windows\system32\ntdll.dll)
0x774715d6 : C:\Windows\system32\ntdll.dll : RtlInitializeExceptionChain

**Comment by luke-jr** [comment]
Ah okay, I was testing with a NF1... Lemme see if I can dig out my NF2 prototype. :|

**Comment by luke-jr** [comment]
Fixed in b4e1c2a, thanks!

**Comment by nanofury** [comment]
Awesome! :)

Just tested and confirmed working!

Thanks Luke!
